shark-android/src/test/java/shark/AndroidResourceIdNamesTest.kt

```diff
@@ -0,0 +1,148 @@
+package shark
+
+import com.nhaarman.mockito_kotlin.any
+import com.nhaarman.mockito_kotlin.mock
+import com.nhaarman.mockito_kotlin.verify
+import com.nhaarman.mockito_kotlin.whenever
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import shark.AndroidResourceIdNames.Companion.FIRST_APP_RESOURCE_ID
+import shark.AndroidResourceIdNames.Companion.RESOURCE_ID_TYPE_ITERATOR
+import java.io.File
+
+class AndroidResourceIdNamesTest {
+
+  @get:Rule
+  var testFolder = TemporaryFolder()
+
+  @Before fun setUp() {
+    AndroidResourceIdNames.resetForTests()
+  }
+
+  @After fun tearDown() {
+    AndroidResourceIdNames.resetForTests()
+  }
+
+  @Test fun `saveToMemory call is cached`() {
+    val getResourceTypeName: (Int) -> String? = mock()
+
+    for (i in 0..2) {
+      AndroidResourceIdNames.saveToMemory(
+          getResourceTypeName = getResourceTypeName,
+          getResourceEntryName = { null }
+      )
+    }
+
+    verify(getResourceTypeName).invoke(any())
+  }
+
+  @Test fun `AndroidResourceIdNames available in heap dump when saveToMemory is called`() {
+    AndroidResourceIdNames.saveToMemory(
+        getResourceTypeName = { null }, getResourceEntryName = { null })
+
+    dumpAndReadHeap { resIdNames ->
+      assertThat(resIdNames).isNotNull
+    }
+  }
+
+  @Test fun `AndroidResourceIdNames not available in heap dump when saveToMemory is not called`() {
+    dumpAndReadHeap { resIdNames ->
+      assertThat(resIdNames).isNull()
+    }
+  }
+
+
+  @Test fun `saveToMemory stores and retrieves resource id`() {
+    val firstIdResourceId = FIRST_APP_RESOURCE_ID
+
+    val getResourceTypeName =
+      createGetResourceTypeName(mapOf(firstIdResourceId to "id"))
+
+    val getResourceEntryName =
+      createGetResourceEntryName(mapOf(FIRST_APP_RESOURCE_ID to "view_container"))
+
+    AndroidResourceIdNames.saveToMemory(getResourceTypeName, getResourceEntryName)
+
+    dumpAndReadHeap { resIdNames ->
+      assertThat(resIdNames!![firstIdResourceId]).isEqualTo("view_container")
+    }
+  }
+
+  @Test fun `id type starts after layout`() {
+    val layoutResourceId = FIRST_APP_RESOURCE_ID
+    val firstIdResourceId = layoutResourceId + RESOURCE_ID_TYPE_ITERATOR
+
+    val getResourceTypeName =
+      createGetResourceTypeName(
+          mapOf(
+              layoutResourceId to "layout",
+              firstIdResourceId to "id"
+          )
+      )
+    val getResourceEntryName =
+      createGetResourceEntryName(mapOf(firstIdResourceId to "view_container"))
+
+    AndroidResourceIdNames.saveToMemory(getResourceTypeName, getResourceEntryName)
+
+    dumpAndReadHeap { resIdNames ->
+      assertThat(resIdNames!![firstIdResourceId]).isEqualTo("view_container")
+    }
+  }
```

Comment:
Really excited about those tests, which run in a normal JVM (no Android SDK dependency) but involve a real JVM heap dump and finding the stored data in memory.